### ISSUE TITLE
TypeError: Cannot read property 'metrics' of undefined page speed

### ISSUE THREAD

**Issue by vishwanath-99** [GoogleChromeLabs/psi#118]
While testing for a page i found that the Module with breaking with a change that happened in the response 

https://developers.google.com/speed/docs/insights/release_notes

As of December 3, 2020, PageSpeed Insights is using the Chrome User Experience Report (CrUX) API to get field data. In the API, this affects the loadingExperience and originLoadingExperience objects.

The PSI module breaks because of the loadingExperience object not defined in the response.

**Comment by JuanMaRuiz** [comment]
Hi @vishwanath-99 could you provide the error trace from the site you are checking? 

I have checked the module using several sites and I have found that `psi` throws an error because some rules has not a detail property

```javascript
rule { id: 'uses-rel-preconnect', weight: 0, group: 'load-opportunities' }
details undefined
TypeError: Cannot read property 'type' of undefined
    at /usr/local/lib/node_modules/psi/lib/output.js:75:20
    at Array.filter (<anonymous>)
    at opportunities (/usr/local/lib/node_modules/psi/lib/output.js:71:34)
    at /usr/local/lib/node_modules/psi/lib/output.js:108:7
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

I don't know if this is related with the change of the API you mention. Maybe @exterkamp could help us with that.

I'm going to make a PR in order fix this unhandled error.

Many thanks for reporting the error.

**Comment by edm00se** [comment]
This looks to be the relevant issue, GoogleChromeLabs/psi#115. It's marked closed as the PR for it is merged, but the merged commit failed their travis tasks, so it appears to be stuck in ci limbo.

**Comment by edm00se** [comment]
@JuanMaRuiz PR #115 may have merged, but it appears the [Travis CI builds are failing](https://travis-ci.org/github/GoogleChromeLabs/psi), so it may be fixed in the code, but it doesn't appear to have published anything new to npm.

**Comment by JuanMaRuiz** [comment]
Hi @edm00se what is the site you are testing with psi and what is the version of psi that you're using?

Best regards.

**Comment by edm00se** [comment]
A few sites I've checked, including:

## 1. the svelte clone of hacker news

```sh
❯ npx psi https://hn.svelte.dev/
npx: installed 269 in 10.428s
TypeError: Cannot read property 'metrics' of undefined
    at /Users/thor/.npm/_npx/99288/lib/node_modules/psi/lib/output.js:104:35
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
~ ❯                                                                                                                         18s
```

## 2. the vue clone of hacker news

```sh
❯ npx psi https://vue-hn.herokuapp.com/
npx: installed 269 in 15.307s
TypeError: Cannot read property 'metrics' of undefined
    at /Users/thor/.npm/_npx/99308/lib/node_modules/psi/lib/output.js:104:35
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
~ ❯                                                                                                                         25s
```

## 3. the NY Times covid tracking page

```sh
❯ npx psi https://www.nytimes.com/interactive/2020/world/coronavirus-maps.html
npx: installed 269 in 13.847s
TypeError: Cannot read property 'type' of undefined
    at /Users/thor/.npm/_npx/99494/lib/node_modules/psi/lib/output.js:73:20
    at Array.filter (<anonymous>)
    at opportunities (/Users/thor/.npm/_npx/99494/lib/node_modules/psi/lib/output.js:71:34)
    at /Users/thor/.npm/_npx/99494/lib/node_modules/psi/lib/output.js:106:7
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
~ ❯                                                                                                                         39s
```

## 4. it also fails for a couple sites that use psi as a GitHub Action in a private source repo I can't link to



As for verison, it's the latest published to npm, when executed from a directory without `node_modules` (as was the case, otherwise it uses the locally installed version). It's how `npx` works, hence my use of it. This is why I then used it with this repo specified, to make use of the latest in the default branch, in case there was a difference between the latest published to npm versus what's in the source; hence my comment on the fact that this repo's Travis CI builds are failing.

**Comment by rndm-user** [comment]
`psi --version` is `4.0.1`

`psi https://nelsondesigncollective.com/`
```
TypeError: Cannot read property 'metrics' of undefined
    at /home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:104:35
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

`psi https://wordpress.com/`
```
TypeError: Cannot read property 'type' of undefined
    at /home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:73:20
    at Array.filter (<anonymous>)
    at opportunities (/home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:71:34)
    at /home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:106:7
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

`psi https://microsoft.com/`
```
TypeError: Cannot read property 'type' of undefined
    at /home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:73:20
    at Array.filter (<anonymous>)
    at opportunities (/home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:71:34)
    at /home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/lib/output.js:106:7
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

`psi https://github.com/`
```
Error: Lighthouse returned error: This request took too long to render.
    at Gaxios.<anonymous> (/home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/node_modules/gaxios/build/src/gaxios.js:73:27)
    at Generator.next (<anonymous>)
    at fulfilled (/home/joshuajarman/.nvm/versions/node/v14.4.0/lib/node_modules/psi/node_modules/gaxios/build/src/gaxios.js:16:58)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by JuanMaRuiz** [comment]
Thanks **@all** for reporting the issue. This [PR](https://github.com/GoogleChromeLabs/psi/pull/120) should fix the error.

@exterkamp build is failing due the _**Quota exceeded**_ error. Could you please take a look to the PR and relaunch the travis job?

Cheers.

**Comment by exterkamp** [comment]
This is actually 2 errors as far as I can tell.

```js
TypeError: Cannot read property 'type' of undefined
```

Is because I haven't updated npm past 4.0.1, which contains the fixes in #116 which adds the `details` validity check.

```js
TypeError: Cannot read property 'metrics' of undefined
```

Is because of the `loadingExperience` object being undefined. Which #120 should fix.

So I think we will need to land #120 and then release a new npm version (v4.1.0?) to push out this fix and #116, which is long overdue.

**Comment by JuanMaRuiz** [comment]
You are right @exterkamp there are two different errors in this issue. But both errors should be fixed with the **PR** #120 see [my comment](https://github.com/GoogleChromeLabs/psi/pull/120#discussion_r553159352) about it.

Once #120 is merged we could release the new version of psi.

Cheers.

**Comment by exterkamp** [comment]
This should be fixed at HEAD, and I will release an npm version soon 😄
